File: frontend/src/services/clientService.js
import api, { API_URLS } from './api';

const BASE_URL = `${API_URLS.CLIENT_SERVICE}/api/clients`;

export const clientService = {
  // Récupérer tous les clients
  getAllClients: async () => {
    const response = await api.get(BASE_URL);
    return response.data;
  },

  // Récupérer un client par ID
  getClientById: async (id) => {
    const response = await api.get(`${BASE_URL}/${id}`);
    return response.data;
  },

  // Récupérer un client par email
  getClientByEmail: async (email) => {
    const response = await api.get(`${BASE_URL}/email/${email}`);
    return response.data;
  },

  // Créer un client
  createClient: async (clientData) => {
    const response = await api.post(BASE_URL, clientData);
    return response.data;
  },

  // Modifier un client
  updateClient: async (id, clientData) => {
    const response = await api.put(`${BASE_URL}/${id}`, clientData);
    return response.data;
  },

  // Supprimer un client
  deleteClient: async (id) => {
    const response = await api.delete(`${BASE_URL}/${id}`);
    return response.data;
  },
};

File: client-service/src/main/java/net/naour/clientservice/services/ClientService.java
package net.naour.clientservice.services;


import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import net.naour.clientservice.entities.*;
import net.naour.clientservice.repositories.ClientRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ClientService {
    @Autowired
    private final ClientRepository clientRepository;

    public List<Client> getAllClients() {
        return clientRepository.findAll();
    }

    public Optional<Client> getClientById(Long id) {
        return clientRepository.findById(id);
    }

    public List<Client> getClientByEmail(String email) {
        return clientRepository.findByEmail(email);
    }

    public List<Client> searchClientsByName(String nom) {
        return clientRepository.findByNomContainingIgnoreCase(nom);
    }

    public List<Client> getClientsByCity(String ville) {
        return clientRepository.findByVille(ville);
    }

    @Transactional
    public Client createClient(Client client) {
        if (clientRepository.existsByEmail(client.getEmail())) {
            throw new RuntimeException("Un client avec cet email existe déjà");
        }
        return clientRepository.save(client);
    }

    @Transactional
    public Client updateClient(Long id, Client clientDetails) {
        Client client = clientRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Client non trouvé avec l'ID: " + id));

        if (!client.getEmail().equals(clientDetails.getEmail()) &&
                clientRepository.existsByEmail(clientDetails.getEmail())) {
            throw new RuntimeException("Cet email est déjà utilisé par un autre client");
        }

        client.setNom(clientDetails.getNom());
        client.setPrenom(clientDetails.getPrenom());
        client.setEmail(clientDetails.getEmail());
        client.setTelephone(clientDetails.getTelephone());
        client.setAdresse(clientDetails.getAdresse());
        client.setVille(clientDetails.getVille());
        client.setCodePostal(clientDetails.getCodePostal());
        client.setNumeroPermis(clientDetails.getNumeroPermis());

        return clientRepository.save(client);
    }

    @Transactional
    public void deleteClient(Long id) {
        if (!clientRepository.existsById(id)) {
            throw new RuntimeException("Client non trouvé avec l'ID: " + id);
        }
        clientRepository.deleteById(id);
    }

    public boolean existsByEmail(String email) {
        return clientRepository.existsByEmail(email);
    }



}
